### PR TITLE
Initial version of ensureconda written in go.

### PR DESCRIPTION
**Pull request by mariusvniekerk**
Checklist
* [ ] Added a ``news`` entry

### CODE REVIEW

File: go/cmd/ensureconda.go [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+			condaExe, err := evaluateFlagPair(cmd, "conda-exe")
+			if err != nil {
+				panic(err)
+			}
+			noInstall, err := cmd.Flags().GetBool("no-install")
+			if err != nil {
+				panic(err)
+			}
+
+			executable, err := EnsureConda(mamba, micromamba, conda, condaExe, true)
```
Comment (cappachu) [inline go/cmd/ensureconda.go:43]: I understand this is out of scope - do we want to be able to optionally upgrade?

File: go/cmd/ensureconda.go [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+			if executable != "" {
+				fmt.Print(executable)
+				os.Exit(0)
+			}
+			if !noInstall {
+				executable, err = EnsureConda(mamba, micromamba, conda, condaExe, noInstall)
+				if err != nil {
+					er(err)
```
Comment (cappachu) [inline go/cmd/ensureconda.go:51]: `panic(err)`? we can just remove the def of "er"

File: go/cmd/ensureconda.go [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+		return TestSitePath
+	}
+	return appdirs.UserDataDir("ensure-conda", "", "", false)
+}
+
+func EnsureConda(mamba bool, micromamba bool, conda bool, condaExe bool, noInstall bool) (string, error) {
+	var executable string
+	dataDir := sitePath()
+	minMambaVersion, _ := version.NewVersion("0.7.3")
+	minCondaVersion, _ := version.NewVersion("4.8.2")
```
Comment (cappachu) [inline go/cmd/ensureconda.go:76]: let's def consts for the min versions? i.e. "0.7.3", etc

File: go/cmd/ensureconda.go [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+		}
+		if !noInstall {
+			exe, err := InstallMicromamba()
+			if err != nil {
+				return "", err
+			}
+			if valid, _ := microMambaVersionCheck(exe); valid {
+				return exe, nil
+			}
+		}
```
Comment (cappachu) [inline go/cmd/ensureconda.go:101]: hmm, we could use an interface (Install, versionCheck, ResolveExecutable) for each micromamba, condaExe, etc.

File: go/cmd/ensureconda.go [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+//		return err
+//	}
+//
+//	st, _ := os.Stat(dst)
+//	err = os.Chmod(dst, st.Mode()|syscall.S_IXUSR)
+//	if err != nil {
+//		return err
+//	}
+//	return nil
+//}
```
Comment (cappachu) [inline go/cmd/ensureconda.go:184]: remove?

File: go/cmd/ensureconda.go [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+
+// Execute executes the root command.
+func Execute() error {
+	return rootCmd.Execute()
+}
+
+func er(msg interface{}) {
+	fmt.Println("Error:", msg)
+	os.Exit(1)
+}
```
Comment (cappachu) [inline go/cmd/ensureconda.go:194]: remove? replace with panic above

File: go/cmd/install.go [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+	_ = os.MkdirAll(sitePath(), 0700)
+	targetFileName := filepath.Join(sitePath(), exeName)
+	if runtime.GOOS == "windows" {
+		targetFileName = targetFileName + ".exe"
+	}
+	return targetFileName
+}
+
+func InstallMicromamba() (string, error) {
+	url := fmt.Sprintf("https://micromamba.snakepit.net/api/micromamba/%s/latest", PlatformSubdir())
```
Comment (cappachu) [inline go/cmd/install.go:32]: const?

File: go/cmd/install.go [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+}
+
+type AnacondaPkgAttr struct {
+	Subdir       string
+	Version      string
+	Build_number int32
+	Timestamp    uint64
+	Source_url   string
+	Md5          string
+}
```
Comment (cappachu) [inline go/cmd/install.go:43]: nit: CamelCase ? BuildNumber, etc.

File: go/cmd/install.go [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+			return a[i].Timestamp < a[j].Timestamp
+		}
+	}
+}
+func (a AnacondaPkgAttrs) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
+
+func InstallCondaStandalone() (string, error) {
+	// Get the most recent conda-standalone
+	subdir := PlatformSubdir()
+	url := "https://api.anaconda.org/package/anaconda/conda-standalone/files"
```
Comment (cappachu) [inline go/cmd/install.go:76]: const

File: go/cmd/install.go [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+		}
+		if !locked {
+			return errors.New("could not lock")
+		}
+
+		file, err := os.OpenFile(targetFileName, os.O_RDWR|os.O_CREATE|os.O_TRUNC, fileInfo.Mode().Perm())
+		n, cpErr := io.Copy(file, tarReader)
+		if err != nil {
+			return err
+		}
```
Comment (cappachu) [inline go/cmd/install.go:182]: move file open err check above copy

File: go/cmd/install.go [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+type AnacondaPkg struct {
+	Size  uint32
+	Attrs AnacondaPkgAttr
+	Type  string
+}
+
+type AnacondaPkgAttrs []AnacondaPkgAttr
+
+func (a AnacondaPkgAttrs) Len() int { return len(a) }
+func (a AnacondaPkgAttrs) Less(i, j int) bool {
```
Comment (cappachu) [inline go/cmd/install.go:54]: nit: rename i, j -> x, y or a,b  -  makes them a little easier to distinguish for the reader

File: go/cmd/resolve.go [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+	for _, dir := range filepath.SplitList(path) {
+		bad := filepath.Join(".pyenv", "shims")
+		if !strings.Contains(dir, bad) {
+			filteredPaths = append(filteredPaths, dir)
+		}
+	}
+	newPathEnv := filepath.Join(filteredPaths...)
+	os.Setenv("PATH", newPathEnv)
+	return FindExecutable(executableName, versionPredicate)
+}
```
Comment (cappachu) [inline go/cmd/resolve.go:47]: noice!

**Review by cappachu — CHANGES_REQUESTED** [review]
looks goood! very pleased this almost exists!

File: go/cmd/install.go (hunk 11 above)
Comment (mariusvniekerk) [reply go/cmd/install.go:54]: `i`, `j` are the canonical ones for Less

**Review by cappachu — APPROVED** [review]
lgtm!
